contracts/bridge/RateLimiter.sol

```diff
@@ -0,0 +1,264 @@
+        if (
+            allowance.resetTimeMin > 0 &&
+            allowance.lastResetMin <= currentMin - allowance.resetTimeMin
+        ) {
+            allowance.spent = 0;
+            // Resets happen in regular intervals and `lastResetMin` should be aligned to that
+            allowance.lastResetMin =
+                currentMin -
+                ((currentMin - allowance.lastResetMin) %
+                    allowance.resetTimeMin);
+        }
+        return allowance;
+    }
+
+    /**
+     * @notice Checks the allowance for a given token. If the new amount exceeds the allowance, it is not updated and false is returned
+     * otherwise true is returned and the transaction can proceed
+     * @param amount to transfer
+     **/
+    function checkAndUpdateAllowance(address token, uint256 amount)
+        external
+        nonReentrant
+        onlyRole(BRIDGE_ROLE)
+        returns (bool)
+    {
+        Allowance memory allowance = getAllowance(token);
+
+        // Update state
+        // @dev reverts if amount > (2^96 - 1)
+        uint96 newSpent = allowance.spent + uint96(amount);
+
+        // Check overflow
+        require(
+            newSpent > allowance.spent,
+            "overflow detected: newSpent > allowance.spent"
+        );
+
+        // do not proceed. Store the transaction for later
+        if (newSpent >= allowance.amount) {
+            return false;
+        }
+
+        allowance.spent = newSpent;
+        updateAllowance(token, allowance);
+
+        return true;
+    }
+
+    function addToRetryQueue(bytes32 kappa, bytes memory toRetry)
+        external
+        onlyRole(BRIDGE_ROLE)
+    {
+        rateLimited.set(kappa, toRetry);
+    }
+
+    function retryByKappa(bytes32 kappa) external onlyRole(LIMITER_ROLE) {
+        (bytes memory toRetry, ) = rateLimited.get(kappa);
+        (bool success, bytes memory returnData) = BRIDGE_ADDRESS.call(toRetry);
+        require(
+            success,
+            Strings.append("could not call bridge:", _getRevertMsg(returnData))
+        );
+        rateLimited.remove(kappa);
+    }
+
+    function retryCount(uint8 count) external onlyRole(LIMITER_ROLE) {
+        // no issues casting to uint8 here. If length is greater then 255, min is always taken
+        uint8 attempts = uint8(
+            MathUpgradeable.min(uint256(count), rateLimited.length())
+        );
+
+        for (uint8 i = 0; i < attempts; i++) {
+            (bytes32 kappa, bytes memory toRetry, ) = rateLimited.at(i);
+            (bool success, bytes memory returnData) = BRIDGE_ADDRESS.call(
+                toRetry
+            );
+            require(
+                success,
+                Strings.append(
+                    "could not call bridge for kappa: ",
+                    Strings.toHex(kappa),
+                    " reverted with: ",
+                    _getRevertMsg(returnData)
+                )
+            );
+
+            rateLimited.remove(kappa);
```

Comment:
Doing `map.remove()` in the same for loop as `map.at()` will break things. Either do two separate loops for iteration and deletion, or change `Map` to queue and use `peek` and `poll` to get the first element and then delete it.